cgfs.py

```diff
@@ -315,38 +1227,61 @@ def create(self, path, mode):
             file = File(fdata, name=fname)
             file.setattr('st_size', fdata['size'])
             file.setattr('st_mtime', fdata['modification_date'])
-            parent.insert(file)
+
+        parent.insert(file)
 
         file.open(bytes('', 'utf8'))
 
         self.fd += 1
+        self._open_files[self.fd] = file
+
         return self.fd
 
-    def flush(self, path, fh=None):
-        file = self.get_file(path, expect_type=File)
-        res = file.flush()
-        if res is not None:
-            file.id = res['id']
+    def fsync(self, path, _, fh):
+        self._flush_or_fsync(path, fh, 'fsync')
+
+    def flush(self, path, fh):
+        self._flush_or_fsync(path, fh, 'flush')
+
+    def _flush_or_fsync(self, path, fh, todo):
+        with self._lock:
+            if fh is None:  # pragma: no cover
+                file = self.get_file(path, expect_type=SingleFile)
+            else:
+                file = self._open_files[fh]
+            res = getattr(file, todo)()
+            if res is not None:
+                file.id = res['id']
 
     def getattr(self, path, fh=None):
+        with self._lock:
+            return self._getattr(path, fh)
+
+    def _getattr(self, path, fh):
         parts = self.split_path(path)
         file = self.get_file(parts)
 
-        if file.stat is None and len(parts) > 3:
-            submission = self.get_submission(path)
+        if isinstance(file, (TempFile, SpecialFile)):
+            return file.getattr()
 
+        if file.stat is None and len(parts) > 3:
             try:
-                query_path = submission.tld + '/' + '/'.join(parts[3:])
+                submission = self.get_submission(path)
             except CGAPIException as e:
                 handle_cgapi_exception(e)
 
+            query_path = submission.tld + '/' + '/'.join(parts[3:])
+
             if isinstance(file, Directory):
                 query_path += '/'
         else:
             submission = None
             query_path = None
 
-        return file.getattr(submission, query_path)
+        attrs = file.getattr(submission, query_path)
+        if self.fixed and isinstance(file, File):
+            attrs['st_mode'] &= ~0o222
```

Comment:
Shouldn't we use `0o220` here instead?

Reply:
Why?

Reply:
So we don't have to rely on the underlying filesystem to throw an error when another user tries to write to it